org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/nogrammarconstraints/NoGrammarConstraintsCodeAction.java

```diff
@@ -301,45 +333,51 @@ public static TextDocumentEdit createXmlModelEdit(String schemaFileName, String
 		xsdWithXmlModel.endPrologOrPI();
 		xsdWithXmlModel.linefeed();
 
-		String xmlModelInsertText = xsdWithXmlModel.toString();
+		String xmlModelInsertText = (documentElement == null && document.getLastChild() != null ? delimiter : "") + xsdWithXmlModel.toString();
 		Position xmlModelPosition = document.positionAt(beforeTagOffset);
 
 		if (StringUtils.isEmpty(targetNamespace)) {
 			return CodeActionFactory.insertEdit(xmlModelInsertText, xmlModelPosition, document.getTextDocument());
 		}
 
-		StringBuilder xmlNamespaceInsertText = new StringBuilder();
-		xmlNamespaceInsertText.append(" xmlns=\"");
-		xmlNamespaceInsertText.append(targetNamespace);
-		xmlNamespaceInsertText.append("\" ");
+		List<TextEdit> edits = new ArrayList<>(2);
+		edits.add(CodeActionFactory.insertEdit(xmlModelInsertText, xmlModelPosition));
 
-		int afterTagOffset = beforeTagOffset + 1 + documentElement.getTagName().length();
-		Position xmlNamespacePosition = document.positionAt(afterTagOffset);
+		if (documentElement != null) {
+			// insert namespace in root element
+			StringBuilder xmlNamespaceInsertText = new StringBuilder();
+			xmlNamespaceInsertText.append(" xmlns=\"");
+			xmlNamespaceInsertText.append(targetNamespace);
+			xmlNamespaceInsertText.append("\" ");
+
+			int afterTagOffset = beforeTagOffset + 1 + documentElement.getTagName().length();
+			Position xmlNamespacePosition = document.positionAt(afterTagOffset);
+			edits.add(CodeActionFactory.insertEdit(xmlNamespaceInsertText.toString(), xmlNamespacePosition));
+		}

```

Comment:
is there any reason why you don't generate an element for xml_model? IMHO to be consistent, you should generate an element like other  grammar.

Reply:
I think this is a good idea since then we can declare the proper namespace (if we have the schema locally). Should I also do it for `.dtd`?

Reply:
I think we should generate in any case the tag element. In the future we will load the grammar to retrieve the proper tag name.